### PR TITLE
Name service pallet

### LINKED ISSUE #1168: Name Service
Naming Service Pallet

let users register usernames connected to their public key
this lets users transfer funds to others using username instead of keys.

Users can register few names, and set only one as primary.

### PR DESCRIPTION
**Pull request by Apokalip**
## Description
Name Service Pallet

Normal Workflow:
register -> PendingRegister Storage(block wait time) -> accept_register -> UsernameRecords(available to be set as Primary name in PrimaryRecords

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs]: commented code

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs]: comments

File: pallets/name-service/Cargo.toml
Comment (ferrell-code) [inline pallets/name-service/Cargo.toml]: update version before merge

File: pallets/name-service/src/lib.rs
Comment (ferrell-code) [inline pallets/name-service/src/lib.rs]: remove?

File: pallets/name-service/src/lib.rs
Comment (ferrell-code) [inline pallets/name-service/src/lib.rs:151]: can remove if not being used

File: pallets/name-service/src/lib.rs
Comment (ferrell-code) [inline pallets/name-service/src/lib.rs]: plan to use third party nft or use your own scheme?

**Review by ferrell-code — COMMENTED** [review]
first pass

File: pallets/name-service/src/lib.rs
Comment (Apokalip) [reply pallets/name-service/src/lib.rs]: This is for now on hold, as victor wants only the username functionality

Comment (ghzlatarev) [reply pallets/name-service/src/lib.rs]: just remove it then and maybe log an issue

File: runtime/dolphin/Cargo.toml
Comment (ghzlatarev) [inline runtime/dolphin/Cargo.toml]: no-std, runtime-benchmarks, try-runtime features?

File: runtime/dolphin/Cargo.toml
Comment (ghzlatarev) [inline runtime/dolphin/Cargo.toml:1]: actually don't do it on Dolphin, directly on Calamari.
I think we'll just get rid of it.

File: runtime/dolphin/src/lib.rs
Comment (ghzlatarev) [inline runtime/dolphin/src/lib.rs]: todo: weights
you just need to run the all-benchmarks workflow on this branch with calamari-dev

File: runtime/dolphin/src/weights/mod.rs
Comment (ghzlatarev) [inline runtime/dolphin/src/weights/mod.rs]: todo

**Comment by ghzlatarev** [comment]
How do i end2end test this for QA ?

**Comment by Apokalip** [comment]
> How do i end2end test this for QA ?

Will add in readme

File: runtime/dolphin/Cargo.toml
Comment (Apokalip) [reply runtime/dolphin/Cargo.toml:1]: Yep was testing on dolphin will move it

File: pallets/name-service/Cargo.toml
Comment (ghzlatarev) [inline pallets/name-service/Cargo.toml]: inconsistent ""

File: pallets/name-service/README.md
Comment (ghzlatarev) [inline pallets/name-service/README.md]: this history seems wrong

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs]: the weights should use the weights.rs file

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs]: was this the default or did we come up with it somehow ?

Comment (Apokalip) [reply pallets/name-service/src/lib.rs]: Not really this is just some default we came up with for max username length

File: pallets/name-service/src/tests.rs
Comment (ghzlatarev) [inline pallets/name-service/src/tests.rs:178]: can you add a check to re-register a new name after de-registering the first one
I would also assert the storage items values just in case.

Comment (ghzlatarev) [reply pallets/name-service/src/tests.rs:178]: Also canceling and re-registering

File: runtime/calamari/src/lib.rs
Comment (ghzlatarev) [inline runtime/calamari/src/lib.rs]: This is ~75 cents at current kma/usd prices. You sure this is good ?

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs:326]: we should validate the username length here as well before searching

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs:360]: validate here as well.
Or you can simply make the UserName a BoundedVec with size 63

File: pallets/name-service/src/lib.rs
Comment (ghzlatarev) [inline pallets/name-service/src/lib.rs:386]: validate username

File: runtime/calamari/src/lib.rs
Comment (Apokalip) [reply runtime/calamari/src/lib.rs]: Honestly Just put a price, still havent gotten a price direction from Victor. Do you have any opinion on what the price should be or atleast a range?

Comment (ghzlatarev) [reply runtime/calamari/src/lib.rs]: Not really no, what does this cost on other chains ?

Comment (Apokalip) [reply runtime/calamari/src/lib.rs]: For ETH i found this ```Yearly renewals cost $5/year for names that are 5 characters or longer. 4 character names cost $160/year, and 3 character names cost $640/year. Fees are paid in ETH. The ETH/USD exchange rate is set by the Chainlink ETH/USD oracle.```

So rtight now the price is 8.97$ for an year and you can add years when you need to

Comment (Apokalip) [reply runtime/calamari/src/lib.rs]: We can put it on Calamari in this form with 5$, will update with Victor if we want to go the same way with yearly registers for Manta

File: pallets/name-service/src/lib.rs
Comment (Apokalip) [reply pallets/name-service/src/lib.rs:326]: Yeah that should limit storage operations

File: runtime/calamari/src/lib.rs
Comment (ghzlatarev) [reply runtime/calamari/src/lib.rs]: Ok.
but how did you get the `8.97$ for an year` ?

Comment (Apokalip) [reply runtime/calamari/src/lib.rs]: Thats the current price right now(*8.21) (Ethereum Name Service or ENS price) but as I said our pricing should obviously be a lot lower. https://coinmarketcap.com/currencies/ethereum-name-service/

**Review by ghzlatarev — DISMISSED** [review]

File: pallets/name-service/src/lib.rs
Comment (zqhxuyuan) [inline pallets/name-service/src/lib.rs]: Can we add some field into Event?

Comment (Apokalip) [reply pallets/name-service/src/lib.rs]: Yeah I can put the event info inside

File: pallets/name-service/README.md
Comment (zqhxuyuan) [inline pallets/name-service/README.md]: plz use markdown grammar like 1. ...

**Review by ghzlatarev — DISMISSED** [review]

File: pallets/name-service/src/lib.rs
Comment (ferrell-code) [inline pallets/name-service/src/lib.rs:392]: unwrap()?

Comment (ferrell-code) [reply pallets/name-service/src/lib.rs:392]: Ah checked above could be a `match` arm, but fine as is.

Maybe add comment above this one as well

File: pallets/name-service/src/lib.rs
Comment (ferrell-code) [inline pallets/name-service/src/lib.rs]: Why change to lowerecase? 

The above callls still store the uppercase version in storage

Comment (Apokalip) [reply pallets/name-service/src/lib.rs]: Yeah will try to get a better regex check, was reusing some old checks from Quihuans draft

**Review by ferrell-code — APPROVED** [review]
nice

**Review by ghzlatarev — APPROVED** [review]

**Comment by Apokalip** [comment]
https://github.com/Manta-Network/Manta/pull/1178 copied for dco test issues

**Pull request closed without merging**
